Documentation/design/assetgeneration.md

```diff
@@ -0,0 +1,122 @@
+# Installer asset generation based on dependency resolution
+
+## Goal
+
+Define generation of assets of varied types in installer based on a dependency graph.
+
+## Overview
+
+The installer generates assets based on the dependency [graph](./dependency.md). To generate an asset, all its dependencies have to be generated. The installer needs to define how assets declare this dependency and how it resolves the graph when asked to generated specific targets. Targets are chosen assets that need to be generated, the installer generates all assets that the target depends on and writes to disk only the targets, consuming any inputs from disk or state from a previous run.
+
+Each asset is individually responsible for declaring its dependencies. Each asset is also responsible resolving conflicts when combining its input from disk and its state from a previous run. The installer ensures all the dependencies for an asset is generated and provides the asset with latest state to generate its own output.
+
+## State
+
+State is an internal state of the installer. It stores the contents of all the assets that have been generated.
+
+An example of the State:
+
+```go
+// State is map of assets
+type State struct {
+    Root string // hash for the root asset
+
+    Objects map[string]struct{
+        Parents []string // hashes for our parents
+
+        Objects map[string] {
+            Name string // the path on disk.
+            Data []byte
+        }
+    }
+}
+```
+
+## Asset
+
+An asset is the generic representation of work-item for installer that needs to be generated. Each asset defines all the other assets that are required for it to generate itself as dependencies.
+
+The asset would usually follows these steps to generate its output:
+
+1. Load components from disk. (If found, mark them from deletion, we consume our inputs.)
+
+2. Load its components from State object given by installer.
+
+3. If both the sources exist,
+
+    * Both are same: nothing to do use the State object.
+
+    * Both differ: maybe try to merge and mark itself **dirty** or error out.
```

Comment:
I thought we decided that we gonna just throw our hands?

Reply:
That's still an valid outcome.